### PR TITLE
Implement query for lirc programs

### PR DESCRIPTION
**Pull request by seanyoung**
Signed-off-by: Sean Young <sean@mess.org>

### CODE REVIEW

**Comment by seanyoung** [comment]
Fixes #2

File: aya/src/programs/lirc_mode2.rs [hunk 1]
```diff
@@ -78,10 +83,63 @@ impl LircMode2 {
 
         Ok(self.data.link(LircLink::new(prog_fd, lircdev_fd)))
     }
+
+    /// Query lirc device for attached programs
+    pub fn query<T: AsRawFd>(target_fd: T) -> Result<Vec<LircLink>, ProgramError> {
+        let mut prog_ids = vec![0u32; 64];
```
Comment (alessandrod) [inline aya/src/programs/lirc_mode2.rs:89]: Could you move the part that calls `bpf_prog_query` and populates the return ids to a shared  `pub(crate)` util in `aya::programs::query()`?

File: aya/src/programs/lirc_mode2.rs [hunk 2]
```diff
@@ -93,6 +151,17 @@ impl LircLink {
             target_fd: Some(unsafe { dup(target_fd) }),
         }
     }
+
+    pub fn info(&self) -> Result<bpf_prog_info, ProgramError> {
+        if let Some(fd) = self.prog_fd {
```
Comment (alessandrod) [inline aya/src/programs/lirc_mode2.rs:156]: I'm not sure how I feel about exposing `bpf_prog_info` as part of the API. What's your use case? I'm thinking we should maybe return some kind of wrapper type.

**Review by alessandrod — CHANGES_REQUESTED** [review]
Thanks for this! Overall I like this approach and I think it will work for other programs as well.

File: aya/src/programs/lirc_mode2.rs (hunk 1 above)
Comment (seanyoung) [reply aya/src/programs/lirc_mode2.rs]: done

File: aya/src/programs/lirc_mode2.rs (hunk 2 above)
Comment (seanyoung) [reply aya/src/programs/lirc_mode2.rs:124]: The bpf_prog_info struct has a lot of fields and a lot of them are simple `u64` fields, so the wrapper will be pretty thin.

So we could have a `struct ProgramInfo(bpf_prog_info)` with a load of accessor functions. Would that work?

Comment (alessandrod) [reply aya/src/programs/lirc_mode2.rs:124]: That sounds good yes, and for now I would implement only the accessors that you need. If people need more accessors they'll open an issue or send a PR, we'll learn about more use cases, and see if there's a better API we can provide.

Comment (seanyoung) [reply aya/src/programs/lirc_mode2.rs:122]: Great, thanks. I've done that, let me know what you think

File: aya/src/programs/mod.rs [hunk 3]
```diff
@@ -533,3 +579,24 @@ impl_try_from_program!(
     CgroupSkb,
     LircMode2
 );
+
+pub struct ProgramInfo(bpf_prog_info);
+
+impl ProgramInfo {
+    pub fn name(&self) -> Option<String> {
```
Comment (alessandrod) [inline aya/src/programs/mod.rs:586]: Shouldn't this return Option<&[u8]>? It's valid for the name not to be utf8, I think callers should (optionally) do the conversion here

File: aya/src/programs/mod.rs [hunk 4]
```diff
@@ -533,3 +579,24 @@ impl_try_from_program!(
     CgroupSkb,
     LircMode2
 );
+
+pub struct ProgramInfo(bpf_prog_info);
```
Comment (alessandrod) [inline aya/src/programs/mod.rs:583]: Can you add some docs for the struct and the accessors please?

File: aya/src/programs/lirc_mode2.rs (hunk 2 above)
Comment (alessandrod) [reply aya/src/programs/lirc_mode2.rs:122]: Looks good thanks!

File: aya/src/programs/mod.rs (hunk 3 above)
Comment (alessandrod) [reply aya/src/programs/mod.rs:586]: > Shouldn't this return &[u8]? It's valid for the name not to be utf8, I think callers should (optionally) do the conversion here

I just realised that we use String for program names everywhere else so I'm leaving this up to you, I'm ok with either

Comment (seanyoung) [reply aya/src/programs/mod.rs]: So we have both of them now as an option: fn name() which gives you `&[u8]` and fn name_as_str() which returns `Option<&str>`.

File: aya/src/programs/mod.rs (hunk 4 above)
Comment (seanyoung) [reply aya/src/programs/mod.rs:584]: I've added some rustdoc (a bit rudimentary)

**Review by alessandrod — APPROVED** [review]

**Comment by alessandrod** [comment]
thank you!

**Pull request merged**